frame/support/test/tests/storage_layers.rs

```diff
@@ -0,0 +1,197 @@
+			Value::set(99);
+			Map::insert("val0", 99);
+			assert_eq!(Value::get(), 99);
+			assert_eq!(Map::get("val0"), 99);
+			Ok(())
+		}));
+
+		assert_eq!(Value::get(), 99);
+		assert_eq!(Map::get("val0"), 99);
+	});
+}
+
+#[test]
+fn storage_transaction_basic_rollback() {
+	TestExternalities::default().execute_with(|| {
+		assert_eq!(Value::get(), 0);
+		assert_eq!(Map::get("val0"), 0);
+
+		assert_noop!(
+			with_storage_layer(|| -> DispatchResult {
+				Value::set(99);
+				Map::insert("val0", 99);
+				assert_eq!(Value::get(), 99);
+				assert_eq!(Map::get("val0"), 99);
+				Err("revert".into())
+			}),
+			"revert"
+		);
+
+		assert_eq!(Value::get(), 0);
+		assert_eq!(Map::get("val0"), 0);
+	});
+}
+
+#[test]
+fn storage_transaction_rollback_then_commit() {
+	TestExternalities::default().execute_with(|| {
+		Value::set(1);
+		Map::insert("val1", 1);
+
+		assert_ok!(with_storage_layer(|| -> DispatchResult {
+			Value::set(2);
+			Map::insert("val1", 2);
+			Map::insert("val2", 2);
+
+			assert_noop!(
+				with_storage_layer(|| -> DispatchResult {
+					Value::set(3);
+					Map::insert("val1", 3);
+					Map::insert("val2", 3);
+					Map::insert("val3", 3);
+
+					assert_eq!(Value::get(), 3);
+					assert_eq!(Map::get("val1"), 3);
+					assert_eq!(Map::get("val2"), 3);
+					assert_eq!(Map::get("val3"), 3);
+
+					Err("revert".into())
+				}),
+				"revert"
+			);
+
+			assert_eq!(Value::get(), 2);
+			assert_eq!(Map::get("val1"), 2);
+			assert_eq!(Map::get("val2"), 2);
+			assert_eq!(Map::get("val3"), 0);
+
+			Ok(())
+		}));
+
+		assert_eq!(Value::get(), 2);
+		assert_eq!(Map::get("val1"), 2);
+		assert_eq!(Map::get("val2"), 2);
+		assert_eq!(Map::get("val3"), 0);
+	});
+}
+
+#[test]
+fn storage_transaction_commit_then_rollback() {
+	TestExternalities::default().execute_with(|| {
+		Value::set(1);
+		Map::insert("val1", 1);
+
+		assert_noop!(
+			with_storage_layer(|| -> DispatchResult {
+				Value::set(2);
+				Map::insert("val1", 2);
+				Map::insert("val2", 2);
+
+				assert_ok!(with_storage_layer(|| -> DispatchResult {
+					Value::set(3);
+					Map::insert("val1", 3);
+					Map::insert("val2", 3);
+					Map::insert("val3", 3);
+
+					assert_eq!(Value::get(), 3);
+					assert_eq!(Map::get("val1"), 3);
+					assert_eq!(Map::get("val2"), 3);
+					assert_eq!(Map::get("val3"), 3);
+
+					Ok(())
+				}));
+
+				assert_eq!(Value::get(), 3);
+				assert_eq!(Map::get("val1"), 3);
+				assert_eq!(Map::get("val2"), 3);
+				assert_eq!(Map::get("val3"), 3);
+
+				Err("revert".into())
+			}),
+			"revert"
+		);
+
+		assert_eq!(Value::get(), 1);
+		assert_eq!(Map::get("val1"), 1);
+		assert_eq!(Map::get("val2"), 0);
+		assert_eq!(Map::get("val3"), 0);
+	});
+}
+
+#[test]
+fn storage_layer_in_decl_module() {
+	TestExternalities::default().execute_with(|| {
+		let origin = 0;
+		assert_ok!(<Module<Runtime>>::value_commits(origin, 2));
+		assert_eq!(Value::get(), 2);
+
+		assert_noop!(<Module<Runtime>>::value_rollbacks(origin, 3), "nah");
```

Comment:
This one fails to revert and sets the new value instead.

Reply:
Yea but that assert failed: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1570632#L1202

Reply:
yeah fixed it

Reply:
this is the issue i want to solve still... calling the function directly, you dont see the storage layer behavior, but if you call `dispatch` you do

Reply:
probably just needs more macro magic